validate_channels and validate_locales executed on GET /api/v2/recipe/

When I do:
```
curl http://localhost:8000/api/v2/recipe/\?ordering\=-last_updated\&page\=1
```
I see the following SQL select statements:

<details>
<summary>All SQL needed for /api/v2/recipe/</summary>
2018-10-18 13:37:20.602 EDT [60025] LOG:  statement: SET TIME ZONE 'UTC'
2018-10-18 13:37:20.603 EDT [60025] LOG:  statement: SELECT COUNT(*) AS "__count" FROM "recipes_recipe"
2018-10-18 13:37:20.606 EDT [60025] LOG:  statement: SELECT "recipes_recipe"."id", "recipes_recipe"."latest_revision_id", "recipes_recipe"."approved_revision_id", "recipes_recipe"."signature_id", "recipes_reciperevision"."id", "recipes_reciperevision"."parent_id", "recipes_reciperevision"."recipe_id", "recipes_reciperevision"."created", "recipes_reciperevision"."updated", "recipes_reciperevision"."user_id", "recipes_reciperevision"."name", "recipes_reciperevision"."action_id", "recipes_reciperevision"."arguments_json", "recipes_reciperevision"."extra_filter_expression", "recipes_reciperevision"."filter_object_json", "recipes_reciperevision"."identicon_seed", "recipes_reciperevision"."enabled_state_id", "recipes_reciperevision"."comment", "recipes_reciperevision"."bug_number", "recipes_action"."id", "recipes_action"."name", "recipes_action"."implementation", "recipes_action"."implementation_hash", "recipes_action"."arguments_schema_json", "recipes_action"."signature_id", "recipes_approvalrequest"."id", "recipes_approvalrequest"."revision_id", "recipes_approvalrequest"."created", "recipes_approvalrequest"."creator_id", "recipes_approvalrequest"."approved", "recipes_approvalrequest"."approver_id", "recipes_approvalrequest"."comment" FROM "recipes_recipe" LEFT OUTER JOIN "recipes_reciperevision" ON ("recipes_recipe"."latest_revision_id" = "recipes_reciperevision"."id") LEFT OUTER JOIN "recipes_action" ON ("recipes_reciperevision"."action_id" = "recipes_action"."id") LEFT OUTER JOIN "recipes_approvalrequest" ON ("recipes_reciperevision"."id" = "recipes_approvalrequest"."revision_id") ORDER BY "recipes_reciperevision"."updated" DESC LIMIT 10
2018-10-18 13:37:20.616 EDT [60025] LOG:  statement: SELECT ("recipes_reciperevision_channels"."reciperevision_id") AS "_prefetch_related_val_reciperevision_id", "recipes_channel"."id", "recipes_channel"."slug", "recipes_channel"."name" FROM "recipes_channel" INNER JOIN "recipes_reciperevision_channels" ON ("recipes_channel"."id" = "recipes_reciperevision_channels"."channel_id") WHERE "recipes_reciperevision_channels"."reciperevision_id" IN (4, 102, 40, 106, 11, 108, 113, 114, 18, 52) ORDER BY "recipes_channel"."slug" ASC
2018-10-18 13:37:20.621 EDT [60025] LOG:  statement: SELECT ("recipes_reciperevision_countries"."reciperevision_id") AS "_prefetch_related_val_reciperevision_id", "recipes_country"."id", "recipes_country"."code", "recipes_country"."name" FROM "recipes_country" INNER JOIN "recipes_reciperevision_countries" ON ("recipes_country"."id" = "recipes_reciperevision_countries"."country_id") WHERE "recipes_reciperevision_countries"."reciperevision_id" IN (4, 102, 40, 106, 11, 108, 113, 114, 18, 52) ORDER BY "recipes_country"."name" ASC
2018-10-18 13:37:20.626 EDT [60025] LOG:  statement: SELECT ("recipes_reciperevision_locales"."reciperevision_id") AS "_prefetch_related_val_reciperevision_id", "recipes_locale"."id", "recipes_locale"."code", "recipes_locale"."name" FROM "recipes_locale" INNER JOIN "recipes_reciperevision_locales" ON ("recipes_locale"."id" = "recipes_reciperevision_locales"."locale_id") WHERE "recipes_reciperevision_locales"."reciperevision_id" IN (4, 102, 40, 106, 11, 108, 113, 114, 18, 52) ORDER BY "recipes_locale"."name" ASC
2018-10-18 13:37:20.632 EDT [60025] LOG:  statement: SELECT (1) AS "a" FROM "recipes_locale" WHERE "recipes_locale"."code" = 'sv-SE' LIMIT 1
2018-10-18 13:37:20.634 EDT [60025] LOG:  statement: SELECT (1) AS "a" FROM "recipes_country" WHERE "recipes_country"."code" = 'SE' LIMIT 1
2018-10-18 13:37:20.635 EDT [60025] LOG:  statement: SELECT (1) AS "a" FROM "recipes_channel" WHERE "recipes_channel"."slug" = 'beta' LIMIT 1
2018-10-18 13:37:20.635 EDT [60025] LOG:  statement: SELECT (1) AS "a" FROM "recipes_channel" WHERE "recipes_channel"."slug" = 'nightly' LIMIT 1
2018-10-18 13:37:20.638 EDT [60025] LOG:  statement: SELECT "recipes_recipe"."id", "recipes_recipe"."latest_revision_id", "recipes_recipe"."approved_revision_id", "recipes_recipe"."signature_id" FROM "recipes_recipe" WHERE "recipes_recipe"."id" = 12
2018-10-18 13:37:20.640 EDT [60025] LOG:  statement: SELECT (1) AS "a" FROM "recipes_locale" WHERE "recipes_locale"."code" = 'sv-SE' LIMIT 1
2018-10-18 13:37:20.641 EDT [60025] LOG:  statement: SELECT (1) AS "a" FROM "recipes_country" WHERE "recipes_country"."code" = 'SE' LIMIT 1
2018-10-18 13:37:20.642 EDT [60025] LOG:  statement: SELECT (1) AS "a" FROM "recipes_channel" WHERE "recipes_channel"."slug" = 'beta' LIMIT 1
2018-10-18 13:37:20.643 EDT [60025] LOG:  statement: SELECT (1) AS "a" FROM "recipes_channel" WHERE "recipes_channel"."slug" = 'nightly' LIMIT 1
2018-10-18 13:37:20.645 EDT [60025] LOG:  statement: SELECT "auth_user"."id", "auth_user"."password", "auth_user"."last_login", "auth_user"."is_superuser", "auth_user"."username", "auth_user"."first_name", "auth_user"."last_name", "auth_user"."email", "auth_user"."is_staff", "auth_user"."is_active", "auth_user"."date_joined" FROM "auth_user" WHERE "auth_user"."id" = 4
2018-10-18 13:37:20.647 EDT [60025] LOG:  statement: SELECT "recipes_reciperevision"."id", "recipes_reciperevision"."parent_id", "recipes_reciperevision"."recipe_id", "recipes_reciperevision"."created", "recipes_reciperevision"."updated", "recipes_reciperevision"."user_id", "recipes_reciperevision"."name", "recipes_reciperevision"."action_id", "recipes_reciperevision"."arguments_json", "recipes_reciperevision"."extra_filter_expression", "recipes_reciperevision"."filter_object_json", "recipes_reciperevision"."identicon_seed", "recipes_reciperevision"."enabled_state_id", "recipes_reciperevision"."comment", "recipes_reciperevision"."bug_number" FROM "recipes_reciperevision" WHERE "recipes_reciperevision"."id" = 110
2018-10-18 13:37:20.648 EDT [60025] LOG:  statement: SELECT "recipes_action"."id", "recipes_action"."name", "recipes_action"."implementation", "recipes_action"."implementation_hash", "recipes_action"."arguments_schema_json", "recipes_action"."signature_id" FROM "recipes_action" WHERE "recipes_action"."id" = 1
2018-10-18 13:37:20.650 EDT [60025] LOG:  statement: SELECT "recipes_approvalrequest"."id", "recipes_approvalrequest"."revision_id", "recipes_approvalrequest"."created", "recipes_approvalrequest"."creator_id", "recipes_approvalrequest"."approved", "recipes_approvalrequest"."approver_id", "recipes_approvalrequest"."comment" FROM "recipes_approvalrequest" WHERE "recipes_approvalrequest"."revision_id" = 110
2018-10-18 13:37:20.652 EDT [60025] LOG:  statement: SELECT "auth_user"."id", "auth_user"."password", "auth_user"."last_login", "auth_user"."is_superuser", "auth_user"."username", "auth_user"."first_name", "auth_user"."last_name", "auth_user"."email", "auth_user"."is_staff", "auth_user"."is_active", "auth_user"."date_joined" FROM "auth_user" WHERE "auth_user"."id" = 4
2018-10-18 13:37:20.653 EDT [60025] LOG:  statement: SELECT "auth_user"."id", "auth_user"."password", "auth_user"."last_login", "auth_user"."is_superuser", "auth_user"."username", "auth_user"."first_name", "auth_user"."last_name", "auth_user"."email", "auth_user"."is_staff", "auth_user"."is_active", "auth_user"."date_joined" FROM "auth_user" WHERE "auth_user"."id" = 4
2018-10-18 13:37:20.653 EDT [60025] LOG:  statement: SELECT "recipes_recipe"."id", "recipes_recipe"."latest_revision_id", "recipes_recipe"."approved_revision_id", "recipes_recipe"."signature_id" FROM "recipes_recipe" WHERE "recipes_recipe"."id" = 20
2018-10-18 13:37:20.656 EDT [60025] LOG:  statement: SELECT "recipes_enabledstate"."id", "recipes_enabledstate"."revision_id", "recipes_enabledstate"."created", "recipes_enabledstate"."creator_id", "recipes_enabledstate"."enabled", "recipes_enabledstate"."carryover_from_id" FROM "recipes_enabledstate" WHERE "recipes_enabledstate"."id" = 1
2018-10-18 13:37:20.657 EDT [60025] LOG:  statement: SELECT COUNT(*) AS "__count" FROM "recipes_locale" INNER JOIN "recipes_reciperevision_locales" ON ("recipes_locale"."id" = "recipes_reciperevision_locales"."locale_id") WHERE "recipes_reciperevision_locales"."reciperevision_id" = 110
2018-10-18 13:37:20.658 EDT [60025] LOG:  statement: SELECT COUNT(*) AS "__count" FROM "recipes_country" INNER JOIN "recipes_reciperevision_countries" ON ("recipes_country"."id" = "recipes_reciperevision_countries"."country_id") WHERE "recipes_reciperevision_countries"."reciperevision_id" = 110
2018-10-18 13:37:20.659 EDT [60025] LOG:  statement: SELECT COUNT(*) AS "__count" FROM "recipes_channel" INNER JOIN "recipes_reciperevision_channels" ON ("recipes_channel"."id" = "recipes_reciperevision_channels"."channel_id") WHERE "recipes_reciperevision_channels"."reciperevision_id" = 110
2018-10-18 13:37:20.659 EDT [60025] LOG:  statement: SELECT (1) AS "a" FROM "recipes_channel" WHERE "recipes_channel"."slug" = 'beta' LIMIT 1
2018-10-18 13:37:20.660 EDT [60025] LOG:  statement: SELECT "auth_user"."id", "auth_user"."password", "auth_user"."last_login", "auth_user"."is_superuser", "auth_user"."username", "auth_user"."first_name", "auth_user"."last_name", "auth_user"."email", "auth_user"."is_staff", "auth_user"."is_active", "auth_user"."date_joined" FROM "auth_user" WHERE "auth_user"."id" = 4
2018-10-18 13:37:20.661 EDT [60025] LOG:  statement: SELECT COUNT(*) AS "__count" FROM "recipes_locale" INNER JOIN "recipes_reciperevision_locales" ON ("recipes_locale"."id" = "recipes_reciperevision_locales"."locale_id") WHERE "recipes_reciperevision_locales"."reciperevision_id" = 110
2018-10-18 13:37:20.662 EDT [60025] LOG:  statement: SELECT COUNT(*) AS "__count" FROM "recipes_country" INNER JOIN "recipes_reciperevision_countries" ON ("recipes_country"."id" = "recipes_reciperevision_countries"."country_id") WHERE "recipes_reciperevision_countries"."reciperevision_id" = 110
2018-10-18 13:37:20.663 EDT [60025] LOG:  statement: SELECT COUNT(*) AS "__count" FROM "recipes_channel" INNER JOIN "recipes_reciperevision_channels" ON ("recipes_channel"."id" = "recipes_reciperevision_channels"."channel_id") WHERE "recipes_reciperevision_channels"."reciperevision_id" = 110
2018-10-18 13:37:20.664 EDT [60025] LOG:  statement: SELECT (1) AS "a" FROM "recipes_channel" WHERE "recipes_channel"."slug" = 'beta' LIMIT 1
2018-10-18 13:37:20.665 EDT [60025] LOG:  statement: SELECT "recipes_recipe"."id", "recipes_recipe"."latest_revision_id", "recipes_recipe"."approved_revision_id", "recipes_recipe"."signature_id" FROM "recipes_recipe" WHERE "recipes_recipe"."id" = 20
2018-10-18 13:37:20.668 EDT [60025] LOG:  statement: SELECT "auth_user"."id", "auth_user"."password", "auth_user"."last_login", "auth_user"."is_superuser", "auth_user"."username", "auth_user"."first_name", "auth_user"."last_name", "auth_user"."email", "auth_user"."is_staff", "auth_user"."is_active", "auth_user"."date_joined" FROM "auth_user" WHERE "auth_user"."id" = 4
2018-10-18 13:37:20.670 EDT [60025] LOG:  statement: SELECT (1) AS "a" FROM "recipes_channel" WHERE "recipes_channel"."slug" = 'beta' LIMIT 1
2018-10-18 13:37:20.671 EDT [60025] LOG:  statement: SELECT "recipes_recipe"."id", "recipes_recipe"."latest_revision_id", "recipes_recipe"."approved_revision_id", "recipes_recipe"."signature_id" FROM "recipes_recipe" WHERE "recipes_recipe"."id" = 19
2018-10-18 13:37:20.673 EDT [60025] LOG:  statement: SELECT (1) AS "a" FROM "recipes_channel" WHERE "recipes_channel"."slug" = 'beta' LIMIT 1
2018-10-18 13:37:20.674 EDT [60025] LOG:  statement: SELECT "auth_user"."id", "auth_user"."password", "auth_user"."last_login", "auth_user"."is_superuser", "auth_user"."username", "auth_user"."first_name", "auth_user"."last_name", "auth_user"."email", "auth_user"."is_staff", "auth_user"."is_active", "auth_user"."date_joined" FROM "auth_user" WHERE "auth_user"."id" = 4
2018-10-18 13:37:20.676 EDT [60025] LOG:  statement: SELECT (1) AS "a" FROM "recipes_country" WHERE "recipes_country"."code" = 'AS' LIMIT 1
2018-10-18 13:37:20.677 EDT [60025] LOG:  statement: SELECT (1) AS "a" FROM "recipes_country" WHERE "recipes_country"."code" = 'AI' LIMIT 1
2018-10-18 13:37:20.678 EDT [60025] LOG:  statement: SELECT "recipes_recipe"."id", "recipes_recipe"."latest_revision_id", "recipes_recipe"."approved_revision_id", "recipes_recipe"."signature_id" FROM "recipes_recipe" WHERE "recipes_recipe"."id" = 16
2018-10-18 13:37:20.681 EDT [60025] LOG:  statement: SELECT (1) AS "a" FROM "recipes_country" WHERE "recipes_country"."code" = 'AS' LIMIT 1
2018-10-18 13:37:20.681 EDT [60025] LOG:  statement: SELECT (1) AS "a" FROM "recipes_country" WHERE "recipes_country"."code" = 'AI' LIMIT 1
2018-10-18 13:37:20.683 EDT [60025] LOG:  statement: SELECT "auth_user"."id", "auth_user"."password", "auth_user"."last_login", "auth_user"."is_superuser", "auth_user"."username", "auth_user"."first_name", "auth_user"."last_name", "auth_user"."email", "auth_user"."is_staff", "auth_user"."is_active", "auth_user"."date_joined" FROM "auth_user" WHERE "auth_user"."id" = 4
2018-10-18 13:37:20.686 EDT [60025] LOG:  statement: SELECT "recipes_recipe"."id", "recipes_recipe"."latest_revision_id", "recipes_recipe"."approved_revision_id", "recipes_recipe"."signature_id" FROM "recipes_recipe" WHERE "recipes_recipe"."id" = 18
2018-10-18 13:37:20.689 EDT [60025] LOG:  statement: SELECT "auth_user"."id", "auth_user"."password", "auth_user"."last_login", "auth_user"."is_superuser", "auth_user"."username", "auth_user"."first_name", "auth_user"."last_name", "auth_user"."email", "auth_user"."is_staff", "auth_user"."is_active", "auth_user"."date_joined" FROM "auth_user" WHERE "auth_user"."id" = 4
2018-10-18 13:37:20.691 EDT [60025] LOG:  statement: SELECT (1) AS "a" FROM "recipes_country" WHERE "recipes_country"."code" = 'AF' LIMIT 1
2018-10-18 13:37:20.692 EDT [60025] LOG:  statement: SELECT "recipes_recipe"."id", "recipes_recipe"."latest_revision_id", "recipes_recipe"."approved_revision_id", "recipes_recipe"."signature_id" FROM "recipes_recipe" WHERE "recipes_recipe"."id" = 17
2018-10-18 13:37:20.695 EDT [60025] LOG:  statement: SELECT (1) AS "a" FROM "recipes_country" WHERE "recipes_country"."code" = 'AF' LIMIT 1
2018-10-18 13:37:20.696 EDT [60025] LOG:  statement: SELECT "auth_user"."id", "auth_user"."password", "auth_user"."last_login", "auth_user"."is_superuser", "auth_user"."username", "auth_user"."first_name", "auth_user"."last_name", "auth_user"."email", "auth_user"."is_staff", "auth_user"."is_active", "auth_user"."date_joined" FROM "auth_user" WHERE "auth_user"."id" = 4
2018-10-18 13:37:20.697 EDT [60025] LOG:  statement: SELECT "recipes_recipe"."id", "recipes_recipe"."latest_revision_id", "recipes_recipe"."approved_revision_id", "recipes_recipe"."signature_id" FROM "recipes_recipe" WHERE "recipes_recipe"."id" = 11
2018-10-18 13:37:20.699 EDT [60025] LOG:  statement: SELECT "auth_user"."id", "auth_user"."password", "auth_user"."last_login", "auth_user"."is_superuser", "auth_user"."username", "auth_user"."first_name", "auth_user"."last_name", "auth_user"."email", "auth_user"."is_staff", "auth_user"."is_active", "auth_user"."date_joined" FROM "auth_user" WHERE "auth_user"."id" = 4
2018-10-18 13:37:20.701 EDT [60025] LOG:  statement: SELECT "recipes_recipe"."id", "recipes_recipe"."latest_revision_id", "recipes_recipe"."approved_revision_id", "recipes_recipe"."signature_id" FROM "recipes_recipe" WHERE "recipes_recipe"."id" = 15
2018-10-18 13:37:20.703 EDT [60025] LOG:  statement: SELECT "auth_user"."id", "auth_user"."password", "auth_user"."last_login", "auth_user"."is_superuser", "auth_user"."username", "auth_user"."first_name", "auth_user"."last_name", "auth_user"."email", "auth_user"."is_staff", "auth_user"."is_active", "auth_user"."date_joined" FROM "auth_user" WHERE "auth_user"."id" = 4
2018-10-18 13:37:20.705 EDT [60025] LOG:  statement: SELECT "recipes_recipe"."id", "recipes_recipe"."latest_revision_id", "recipes_recipe"."approved_revision_id", "recipes_recipe"."signature_id" FROM "recipes_recipe" WHERE "recipes_recipe"."id" = 10
2018-10-18 13:37:20.708 EDT [60025] LOG:  statement: SELECT "recipes_recipe"."id", "recipes_recipe"."latest_revision_id", "recipes_recipe"."approved_revision_id", "recipes_recipe"."signature_id" FROM "recipes_recipe" WHERE "recipes_recipe"."id" = 14

</details>

That's 55 SELECT statements. 
Potentially bad prefetching or select_related aside, it does execute things like:

```sql
SELECT (1) AS "a" FROM "recipes_channel" WHERE "recipes_channel"."slug" = 'nightly' LIMIT 1;
SELECT (1) AS "a" FROM "recipes_channel" WHERE "recipes_channel"."slug" = 'beta' LIMIT 1;
SELECT (1) AS "a" FROM "recipes_country" WHERE "recipes_country"."code" = 'SE' LIMIT 1;
SELECT (1) AS "a" FROM "recipes_locale" WHERE "recipes_locale"."code" = 'sv-SE' LIMIT 1;
```
etc.

It's a bit weird because I thought things like [`validate_channels`](https://github.com/mozilla/normandy/blob/23f23bea0c4bfb5eabd012f33bd3830e5b536ce3/normandy/recipes/filters.py#L56) where only for when you try to do writes.

Here's the traceback you get if you remove that `assert`'ion:

```
Traceback (most recent call last):
  File "/Users/peterbe/virtualenvs/normandy/lib/python3.6/site-packages/django/core/handlers/exception.py", line 35, in inner
    response = get_response(request)
  File "/Users/peterbe/virtualenvs/normandy/lib/python3.6/site-packages/django/core/handlers/base.py", line 128, in _get_response
    response = self.process_exception_by_middleware(e, request)
  File "/Users/peterbe/virtualenvs/normandy/lib/python3.6/site-packages/django/core/handlers/base.py", line 126, in _get_response
    response = wrapped_callback(request, *callback_args, **callback_kwargs)
  File "/Users/peterbe/virtualenvs/normandy/lib/python3.6/site-packages/django/views/decorators/csrf.py", line 54, in wrapped_view
    return view_func(*args, **kwargs)
  File "/Users/peterbe/virtualenvs/normandy/lib/python3.6/site-packages/rest_framework/viewsets.py", line 103, in view
    return self.dispatch(request, *args, **kwargs)
  File "/Users/peterbe/virtualenvs/normandy/lib/python3.6/site-packages/rest_framework/views.py", line 483, in dispatch
    response = self.handle_exception(exc)
  File "/Users/peterbe/virtualenvs/normandy/lib/python3.6/site-packages/rest_framework/views.py", line 443, in handle_exception
    self.raise_uncaught_exception(exc)
  File "/Users/peterbe/virtualenvs/normandy/lib/python3.6/site-packages/rest_framework/views.py", line 480, in dispatch
    response = handler(request, *args, **kwargs)
  File "/Users/peterbe/virtualenvs/normandy/lib/python3.6/site-packages/django/views/decorators/cache.py", line 31, in _cache_controlled
    response = viewfunc(request, *args, **kw)
  File "/Users/peterbe/dev/MOZILLA/NORMANDY/normandy/normandy/base/api/mixins.py", line 13, in list
    return super().list(request, *args, **kwargs)
  File "/Users/peterbe/virtualenvs/normandy/lib/python3.6/site-packages/rest_framework/mixins.py", line 45, in list
    return self.get_paginated_response(serializer.data)
  File "/Users/peterbe/virtualenvs/normandy/lib/python3.6/site-packages/rest_framework/serializers.py", line 764, in data
    ret = super(ListSerializer, self).data
  File "/Users/peterbe/virtualenvs/normandy/lib/python3.6/site-packages/rest_framework/serializers.py", line 262, in data
    self._data = self.to_representation(self.instance)
  File "/Users/peterbe/virtualenvs/normandy/lib/python3.6/site-packages/rest_framework/serializers.py", line 682, in to_representation
    self.child.to_representation(item) for item in iterable
  File "/Users/peterbe/virtualenvs/normandy/lib/python3.6/site-packages/rest_framework/serializers.py", line 682, in <listcomp>
    self.child.to_representation(item) for item in iterable
  File "/Users/peterbe/virtualenvs/normandy/lib/python3.6/site-packages/rest_framework/serializers.py", line 527, in to_representation
    ret[field.field_name] = field.to_representation(attribute)
  File "/Users/peterbe/virtualenvs/normandy/lib/python3.6/site-packages/rest_framework/serializers.py", line 514, in to_representation
    attribute = field.get_attribute(instance)
  File "/Users/peterbe/virtualenvs/normandy/lib/python3.6/site-packages/rest_framework/fields.py", line 441, in get_attribute
    return get_attribute(instance, self.source_attrs)
  File "/Users/peterbe/virtualenvs/normandy/lib/python3.6/site-packages/rest_framework/fields.py", line 100, in get_attribute
    instance = getattr(instance, attr)
  File "/Users/peterbe/dev/MOZILLA/NORMANDY/normandy/normandy/recipes/models.py", line 376, in filter_expression
    parts.append(filter.to_jexl())
  File "/Users/peterbe/dev/MOZILLA/NORMANDY/normandy/normandy/recipes/filters.py", line 99, in to_jexl
    locales = ",".join(f'"{l}"' for l in self.data["locales"])
  File "/Users/peterbe/virtualenvs/normandy/lib/python3.6/site-packages/rest_framework/serializers.py", line 560, in data
    ret = super(Serializer, self).data
  File "/Users/peterbe/virtualenvs/normandy/lib/python3.6/site-packages/rest_framework/serializers.py", line 258, in data
    raise AssertionError(msg)
AssertionError: When a serializer is passed a `data` keyword argument you must call `.is_valid()` before attempting to access the serialized `.data` representation.
You should either call `.is_valid()` first, or access `.initial_data` instead.
```

--- a/normandy/recipes/filters.py
+++ b/normandy/recipes/filters.py
@@ -63,7 +63,7 @@ class ChannelFilter(BaseFilter):
         return value
 
     def to_jexl(self):
-        channels = ",".join(f'"{c}"' for c in self.data["channels"])
+        channels = ",".join(f'"{c}"' for c in self.initial_data["channels"])
         return f"normandy.channel in [{channels}]"
 
 
@@ -95,7 +95,7 @@ class LocaleFilter(BaseFilter):
         return value
 
     def to_jexl(self):
-        locales = ",".join(f'"{l}"' for l in self.data["locales"])
+        locales = ",".join(f'"{l}"' for l in self.initial_data["locales"])
         return f"normandy.locale in [{locales}]"
 
 
@@ -126,7 +126,7 @@ class CountryFilter(BaseFilter):
         return value
 
     def to_jexl(self):
-        countries = ",".join(f'"{c}"' for c in self.data["countries"])
+        countries = ",".join(f'"{c}"' for c in self.initial_data["countries"])
         return f"normandy.country in [{countries}]"
 
 
@@ -187,10 +187,10 @@ class BucketSampleFilter(BaseFilter):
     input = serializers.ListField(child=serializers.CharField(), min_length=1)
 
     def to_jexl(self):
-        inputs = ",".join(f"{i}" for i in self.data["input"])
-        start = self.data["start"]
-        count = self.data["count"]
-        total = self.data["total"]
+        inputs = ",".join(f"{i}" for i in self.initial_data["input"])
+        start = self.initial_data["start"]
+        count = self.initial_data["count"]
+        total = self.initial_data["total"]
         return f"[{inputs}]|bucketSample({start},{count},{total})"
 
 
@@ -230,8 +230,8 @@ class StableSampleFilter(BaseFilter):
     input = serializers.ListField(child=serializers.CharField(), min_length=1)
 
     def to_jexl(self):
-        inputs = ",".join(f"{i}" for i in self.data["input"])
-        rate = self.data["rate"]
+        inputs = ",".join(f"{i}" for i in self.initial_data["input"])
+        rate = self.initial_data["rate"]
         return f"[{inputs}]|stableSample({rate})"
 
 
@@ -268,7 +268,8 @@ class VersionFilter(BaseFilter):
         #   (normandy.version >= 57 && normandy.version < 58)
 
         return "||".join(
-            f'(normandy.version>="{v}"&&normandy.version<"{v + 1}")' for v in self.data["versions"]
+            f'(normandy.version>="{v}"&&normandy.version<"{v + 1}")'
+            for v in self.initial_data["versions"]
         )
 
 

--- a/normandy/recipes/models.py
+++ b/normandy/recipes/models.py
@@ -372,7 +372,6 @@ class RecipeRevision(DirtyFieldsMixin, models.Model):
 
         for obj in self.filter_object:
             filter = filters.from_data(obj)
-            assert filter.is_valid(), [obj, filter.errors]
             parts.append(filter.to_jexl())
 
         if self.extra_filter_expression:

--- a/normandy/recipes/tests/__init__.py
+++ b/normandy/recipes/tests/__init__.py
@@ -1,4 +1,5 @@
 import hashlib
+import json
 
 from django.utils import timezone
 
@@ -23,6 +24,7 @@ from normandy.recipes.models import (
 class ChannelFactory(factory.DjangoModelFactory):
     class Meta:
         model = Channel
+        django_get_or_create = ("slug",)
 
     slug = "beta"
     name = "Beta"
@@ -31,6 +33,7 @@ class ChannelFactory(factory.DjangoModelFactory):
 class CountryFactory(factory.DjangoModelFactory):
     class Meta:
         model = Country
+        django_get_or_create = ("code",)
 
     code = "SE"
     name = "Sweden"
@@ -39,6 +42,7 @@ class CountryFactory(factory.DjangoModelFactory):
 class LocaleFactory(factory.DjangoModelFactory):
     class Meta:
         model = Locale
+        django_get_or_create = ("code",)
 
     code = "sv"
     name = "Swedish"
@@ -156,6 +160,74 @@ class RecipeRevisionFactory(factory.DjangoModelFactory):
     recipe = factory.SubFactory(RecipeFactory)
     identicon_seed = FuzzyIdenticonSeed()
     comment = FuzzyUnicode()
+    extra_filter_expression = factory.fuzzy.FuzzyChoice(["true", "false"])
+
+    @factory.lazy_attribute
+    def filter_object_json(self):
+        filters = [
+            ChannelFilterFactory(),
+            LocaleFilterFactory(),
+            CountryFilterFactory(),
+            StableSampleFilterFactory(),
+            BucketSampleFilterFactory(),
+        ]
+        return json.dumps(filters)
+
+
+class DictFactory(factory.Factory):
+    class Meta:
+        model = dict
+
+
+class ChannelFilterFactory(DictFactory):
+    class Params:
+        channel_objects = factory.LazyAttribute(lambda o: [ChannelFactory()])
+
+    type = "channel"
+    channels = factory.LazyAttribute(lambda o: [channel.slug for channel in o.channel_objects])
+
+
+class LocaleFilterFactory(DictFactory):
+    class Params:
+        locale_objects = factory.LazyAttribute(lambda o: [LocaleFactory()])
+
+    type = "locale"
+    locales = factory.LazyAttribute(lambda o: [locale.code for locale in o.locale_objects])
+
+
+class CountryFilterFactory(DictFactory):
+    class Params:
+        country_objects = factory.LazyAttribute(lambda o: [CountryFactory()])
+
+    type = "country"
+    countries = factory.LazyAttribute(lambda o: [country.code for country in o.country_objects])
+
+
+class FuzzySampleInputs(factory.fuzzy.FuzzyChoice):
+    def __init__(self, **kwargs):
+        super().__init__(
+            [
+                ["normandy.userId"],
+                ["normandy.userId", "42"],
+                ["normandy.userId", '"some study"'],
+                ["normandy.userId", "normandy.recipe.id"],
+                ["normandy.recipe.id", "normandy.userId"],
+            ]
+        )
+
+
+class StableSampleFilterFactory(DictFactory):
+    type = "stableSample"
+    input = FuzzySampleInputs()
+    rate = factory.fuzzy.FuzzyFloat(0.0, 1.0)
+
+
+class BucketSampleFilterFactory(DictFactory):
+    type = "bucketSample"
+    input = FuzzySampleInputs()
+    total = factory.fuzzy.FuzzyInteger(100, 10000)
+    start = factory.LazyAttribute(lambda o: factory.fuzzy.FuzzyInteger(0, o.total).fuzz())
+    count = factory.LazyAttribute(lambda o: factory.fuzzy.FuzzyInteger(0, o.total).fuzz())
 
 
 class ApprovalRequestFactory(factory.DjangoModelFactory):

--- a/normandy/recipes/tests/api/v2/test_api.py
+++ b/normandy/recipes/tests/api/v2/test_api.py
@@ -416,7 +416,9 @@ class TestRecipeAPI(object):
     @pytest.mark.django_db
     class TestUpdates(object):
         def test_it_can_edit_recipes(self, api_client):
-            recipe = RecipeFactory(name="unchanged", extra_filter_expression="true")
+            recipe = RecipeFactory(
+                name="unchanged", extra_filter_expression="true", filter_object_json=None
+            )
             old_revision_id = recipe.revision_id
 
             res = api_client.patch(
@@ -703,7 +705,7 @@ class TestRecipeAPI(object):
 
             Recipe.objects.get(id=recipe_data["id"])
             assert recipe_data["filter_expression"] == (
-                "([normandy.userId,normandy.recipeId]|bucketSample(1.0,2.0,3.0)) && (true)"
+                "([normandy.userId,normandy.recipeId]|bucketSample(1,2,3)) && (true)"
             )
 
         def test_bucket_sample_correct_fields(self, api_client):
@@ -1556,7 +1558,7 @@ class TestFilterObjects(object):
         )
         assert res.status_code == 201, res.json()
         assert res.json()["filter_expression"] == (
-            "[normandy.userId,normandy.recipeId]|bucketSample(1.0,2.0,3.0)"
+            "[normandy.userId,normandy.recipeId]|bucketSample(1,2,3)"
         )
 
     def test_bucket_sample_correct_fields(self, api_client):

--- a/normandy/recipes/tests/api/v2/test_serializers.py
+++ b/normandy/recipes/tests/api/v2/test_serializers.py
@@ -18,7 +18,7 @@ from normandy.recipes.api.v2.serializers import (
 @pytest.mark.django_db()
 class TestRecipeSerializer:
     def test_it_works(self, rf):
-        recipe = RecipeFactory(arguments={"foo": "bar"})
+        recipe = RecipeFactory(arguments={"foo": "bar"}, filter_object_json=None)
         approval = ApprovalRequestFactory(revision=recipe.latest_revision)
         action = recipe.action
         serializer = RecipeSerializer(recipe, context={"request": rf.get("/")})

--- a/normandy/recipes/tests/api/v3/test_api.py
+++ b/normandy/recipes/tests/api/v3/test_api.py
@@ -464,7 +464,9 @@ class TestRecipeAPI(object):
     @pytest.mark.django_db
     class TestUpdates(object):
         def test_it_can_edit_recipes(self, api_client):
-            recipe = RecipeFactory(name="unchanged", extra_filter_expression="true")
+            recipe = RecipeFactory(
+                name="unchanged", extra_filter_expression="true", filter_object_json=None
+            )
             old_revision_id = recipe.revision_id
 
             res = api_client.patch(
@@ -769,7 +771,7 @@ class TestRecipeAPI(object):
 
             Recipe.objects.get(id=recipe_data["id"])
             assert recipe_data["latest_revision"]["filter_expression"] == (
-                "([normandy.userId,normandy.recipeId]|bucketSample(1.0,2.0,3.0)) && (true)"
+                "([normandy.userId,normandy.recipeId]|bucketSample(1,2,3)) && (true)"
             )
 
         def test_bucket_sample_correct_fields(self, api_client):
@@ -1627,7 +1629,7 @@ class TestFilterObjects(object):
         )
         assert res.status_code == 201, res.json()
         assert res.json()["latest_revision"]["filter_expression"] == (
-            "[normandy.userId,normandy.recipeId]|bucketSample(1.0,2.0,3.0)"
+            "[normandy.userId,normandy.recipeId]|bucketSample(1,2,3)"
         )
 
     def test_bucket_sample_correct_fields(self, api_client):

--- a/normandy/recipes/tests/test_models.py
+++ b/normandy/recipes/tests/test_models.py
@@ -268,10 +268,10 @@ class TestRecipe(object):
         assert recipe.revision_id == revision_id
 
     def test_filter_expression(self):
-        r = RecipeFactory()
+        r = RecipeFactory(extra_filter_expression="", filter_object_json=None)
         assert r.filter_expression == ""
 
-        r = RecipeFactory(extra_filter_expression="2 + 2 == 4")
+        r = RecipeFactory(extra_filter_expression="2 + 2 == 4", filter_object_json=None)
         assert r.filter_expression == "2 + 2 == 4"
 
     def test_canonical_json(self):
@@ -280,6 +280,7 @@ class TestRecipe(object):
             arguments_json='{"foo": 1, "bar": 2}',
             extra_filter_expression="2 + 2 == 4",
             name="canonical",
+            filter_object_json=None,
         )
         # Yes, this is really ugly, but we really do need to compare an exact
         # byte sequence, since this is used for hashing and signing
@@ -403,6 +404,7 @@ class TestRecipe(object):
             action=a1,
             arguments={"message": "something"},
             extra_filter_expression="something !== undefined",
+            filter_object_json=None,
         )
         a2 = ActionFactory()
         recipe.revise(name="changed", action=a2)

